patches/server/1005-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,57 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: TreemanKing <67459602+TreemanKing@users.noreply.github.com>
+Date: Wed, 16 Aug 2023 11:53:44 +1000
+Subject: [PATCH] PreSpawnerSearchEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+index 633500aefd515df5dadda3802b94079f75a03fa0..408d360679e2acafa8fb711b92ec851f7d7d1648 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -46,6 +46,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -57,6 +58,16 @@ public abstract class BaseSpawner {
+     public boolean isNearPlayer(Level world, BlockPos pos) {
+         return world.hasNearbyAlivePlayerThatAffectsSpawning((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - Affects Spawning API
+     }
++    // Paper Start - PreSpawnerSearchEvent (Convert ServerPlayers into BukkitPlayer)
++    public java.util.Collection<org.bukkit.entity.Player> convertServerPlayersToBukkitPlayers(java.util.List<net.minecraft.server.level.ServerPlayer> serverPlayers) {
++        java.util.Collection<org.bukkit.entity.Player> bukkitPlayers = new java.util.ArrayList<>(serverPlayers.size());
++
++        for (net.minecraft.server.level.ServerPlayer human : serverPlayers) {
++            bukkitPlayers.add(human.getBukkitEntity());
++        }
++        return bukkitPlayers;
++    }
++    // Paper End
+ 
+     public void clientTick(Level world, BlockPos pos) {
+         if (!this.isNearPlayer(world, pos)) {
+@@ -86,7 +97,20 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
+-        if (this.isNearPlayer(world, pos)) {
++
++        // Paper start - PreSpawnerSearchEvent
++        if (--searchCooldownTicks > 0) return;
++
++        java.util.Collection<org.bukkit.entity.Player> activatingPlayers = convertServerPlayersToBukkitPlayers(world.getNearbyPlayers(null, (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)); // Uses logic from hasNearbyAlivePlayerThatAffectsSpawning
```

Comment:
This needs a diff on change in MobSpawnerAbstract#isNearPlayer so that changes in there cause the patch to fail, otherwise we might miss changes to said method in an update, leaving us with then faulty copied logic.

Reply:
From where you copied this from (the inside of the `isNearPlayer` method in `MobSpawnerAbstract`).

Reply:
Not really sure how to do this. Is it by annotation or something similiar?

Also, I'm not sure why it reordered the patches the way it did.

Reply:
Not really sure how to do this. Is it by annotation or something similiar?

Also, I'm not sure why it reordered the patches the way it did.

Reply:
You can put a comment "// diff on change" in the isNearPlayer method to be sure to prevent desync even if this method is completely rewritten by paper. Also currently the pr doesn't allow point a and c to happens so see my first comment about that (the edit) but i think the event should be called before the player check to allow a and c.